### PR TITLE
Add a node selector in L2 announcement and BGP announcement

### PR DESCRIPTION
**Pull request by fedepaol**
As described in https://github.com/metallb/metallb/blob/main/design/pool-configuration.md , we add a node selector to both the L2 advertisement and the BGP advertisement.

The semantic is, if the selector is present we should advertise only from the nodes selected by the selector. An empty selector is interpreted as willing to select all the nodes.

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -538,11 +550,18 @@ func bfdProfileFromCR(p metallbv1beta1.BFDProfile) (*BFDProfile, error) {
 	return res, nil
 }
 
-func l2AdvertisementFromCR(crdAd metallbv1beta1.L2Advertisement) *L2Advertisement {
-	return &L2Advertisement{}
+func l2AdvertisementFromCR(crdAd metallbv1beta1.L2Advertisement, nodes []corev1.Node) (*L2Advertisement, error) {
+	selected, err := selectedNodes(nodes, crdAd.Spec.NodeSelectors)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to parse node selector for ls %s", crdAd.Name)
```
Comment (sabinaaledort) [inline internal/config/config.go:556]: what is ls?

Comment (sabinaaledort) [reply internal/config/config.go:556]: label selector?

File: internal/k8s/controllers/config_controller.go [hunk 2]
```diff
@@ -89,6 +89,12 @@ func (r *ConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		return ctrl.Result{}, err
 	}
 
+	var nodes corev1.NodeList
+	if err := r.List(ctx, &nodes); err != nil {
+		level.Error(r.Logger).Log("controller", "ConfigReconciler", "message", "failed to get bgp advertisements", "error", err)
```
Comment (sabinaaledort) [inline internal/k8s/controllers/config_controller.go:94]: failed to get nodes

Comment (fedepaol) [reply internal/k8s/controllers/config_controller.go]: Fixed

File: config/manifests/metallb-frr.yaml [hunk 3]
```diff
@@ -1125,6 +1125,7 @@ rules:
   - ""
   resources:
   - services
+  - nodes
```
Comment (sabinaaledort) [inline config/manifests/metallb-frr.yaml]: is this related to this commit? (add empty line between functions)

File: config/rbac/role.yaml [hunk 4]
```diff
@@ -9,6 +9,7 @@ rules:
       - ""
     resources:
       - services
+      - nodes
```
Comment (sabinaaledort) [inline config/rbac/role.yaml]: is this related to this commit? (add empty line between functions)

**Comment by fedepaol** [comment]
> "Fix helm when memberlist is enabled" commit: double "the" in the commit message

I intended it as "memberlist as a library", so I think it's fine as it is

File: e2etest/bgptests/node_selector.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+		cs = f.ClientSet
+	})
+
+	table.DescribeTable("Two services, two distinct advertisements with different node selectors",
+		func(pairingIPFamily ipfamily.Family, addresses []string, nodesForFirstPool, nodesForSecondPool []int) {
+			var allNodes *corev1.NodeList
+			allNodes, err := cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+
+			expectedNodesForFirstPool := nodesForSelection(allNodes.Items, nodesForFirstPool)
```
Comment (oribon) [inline e2etest/bgptests/node_selector.go:83]: can we use `"k8s.io/kubernetes/test/e2e/framework/node"`'s GetBoundedReadySchedulableNodes func instead? (we should probably leverage that package more anyways)

also, couldn't we get the same nodes for first/second pool (do we care about it here)?

File: charts/metallb/templates/crds.yaml [hunk 6]
```diff
@@ -317,6 +317,8 @@ status:
   conditions: []
   storedVersions: []
 ---
+
+---
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply charts/metallb/templates/crds.yaml:321]: Fixed

File: internal/config/config.go [hunk 7]
```diff
@@ -816,9 +843,36 @@ func validateDuplicateCommunities(communities []string) error {
...
+	labelSelectors := []labels.Selector{}
+	for _, selector := range selectors {
+		l, err := metav1.LabelSelectorAsSelector(&selector)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Invalid label selector %v", selector)
+		}
+		labelSelectors = append(labelSelectors, l)
+	}
+
+	res := make(map[string]bool)
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply internal/config/config.go:863]: Nope, and yes, we are using the map as a set as go lacks of the set (and in this case it's clear enough), no need for custom types.

File: internal/k8s/controllers/config_controller.go [hunk 8]
```diff
@@ -129,6 +136,7 @@ func (r *ConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 func (r *ConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&metallbv1beta2.BGPPeer{}).
+		Watches(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{}).
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply internal/k8s/controllers/config_controller.go:139]: Yes, for the speaker. There are two different callbacks, one for nodes only and one for config that have different purpouses.
On the controller, I think we can remove the node handler because the config handler eventually converges in syncpeers.

File: e2etest/bgptests/node_selector.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+			speakerPods, err := metallb.SpeakerPods(cs)
+			framework.ExpectNoError(err)
+			for _, pod := range speakerPods {
+				if len(pod.Spec.Containers) == 1 { // we dump only in case of frr
+					break
+				}
+				podExec := executor.ForPod(pod.Namespace, pod.Name, "frr")
+				dump, err := frr.RawDump(podExec, "/etc/frr/frr.conf", "/etc/frr/frr.log")
+				framework.Logf("External frr dump for pod %s\n%s %v", pod.Name, dump, err)
+			}
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply e2etest/bgptests/node_selector.go:54]: Or actually, let me wrap it

File: internal/config/config.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (fedepaol) [reply internal/config/config.go:863]: From https://go.dev/blog/maps : "For instance, a map of boolean values can be used as a set-like data structure (recall that the zero value for the boolean type is false). "

File: internal/k8s/controllers/config_controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (fedepaol) [reply internal/k8s/controllers/config_controller.go:139]: Actually, I tried and we _could_ get rid of it by calling node handler in the config reconciler.
However, it feels cleaner to have one reconciler per handler, and the node handler has a different purpouse from the confighandler (setting the labels in the bgp protocol), so unless we add a "config + node handler" where we do both (which seems super ugly) I think it's more clear to leave it as it is unless we realize it brings much performance penalty.

File: e2etest/bgptests/node_selector.go [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+			for _, c := range FRRContainers {
+				err := frrcontainer.PairWithNodes(cs, c, ipfamily.IPv4)
+				framework.ExpectNoError(err)
+			}
+
+			for i, c := range FRRContainers {
+				selected := nodesForSelection(allNodes.Items, nodesForPeers[i])
+				nonSelected := nodesNotSelected(allNodes.Items, nodesForPeers[i])
+				validateFRRPeeredWithNodesByList(cs, selected, c, ipfamily.IPv4)
+				validateFRRNotPeeredWithNodes(cs, nonSelected, c, ipfamily.IPv4)
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply e2etest/bgptests/node_selector.go:239]: We want not only to ensure that the sessoin is established with the subset of nodes, we want also to ensure that it's explicitly not established for the rest. That's why we check for the error "not established"

**Comment by PerilousApricot** [comment]
Awesome! This is a great addition!
